lib/handlers/authorize-handler.js

```diff
@@ -0,0 +1,310 @@
+    this.getAuthorizationCodeLifetime(),
+    this.getClient(request),
+    this.getUser(request, response)
+  ];
+
+  return Promise.all(fns)
+    .bind(this)
+    .spread(function(authorizationCode, expiresAt, client, user) {
+      var scope;
+      var state;
+
+      return Promise.bind(this)
+        .then(function() {
+          scope = this.getScope(request);
+          state = this.getState(request);
+
+          return this.saveAuthorizationCode(authorizationCode, expiresAt, scope, client, user);
+        })
+        .then(function(code) {
+          var responseType = this.getResponseType(request, code);
+          var redirectUri = this.buildSuccessRedirectUri(client.redirectUri, responseType);
+
+          this.updateResponse(response, redirectUri, state);
+
+          return code;
+        })
+        .catch(function(e) {
+          if (!(e instanceof OAuthError)) {
+            e = new ServerError(e);
+          }
+
+          var redirectUri = this.buildErrorRedirectUri(client.redirectUri, e);
+
+          this.updateResponse(response, redirectUri, state);
+
+          throw e;
+        });
+    });
+};
+
+/**
+ * Generate authorization code.
+ */
+
+AuthorizeHandler.prototype.generateAuthorizationCode = function() {
+  if (this.model.generateAuthorizationCode) {
+    return Promise.try(this.model.generateAuthorizationCode);
+  }
+
+  return tokenUtil.generateRandomToken();
+};
+
+/**
+ * Get authorization code lifetime.
+ */
+
+AuthorizeHandler.prototype.getAuthorizationCodeLifetime = function() {
+  var expires = new Date();
+
+  expires.setSeconds(expires.getSeconds() + this.authorizationCodeLifetime);
+
+  return expires;
+};
+
+/**
+ * Get the client from the model.
+ */
+
+AuthorizeHandler.prototype.getClient = function(request) {
+  var clientId = request.body.client_id || request.query.client_id;
+
+  if (!clientId) {
+    throw new InvalidRequestError('Missing parameter: `client_id`');
+  }
+
+  if (!is.vschar(clientId)) {
+    throw new InvalidRequestError('Invalid parameter: `client_id`');
+  }
+
+  var redirectUri = request.body.redirect_uri || request.query.redirect_uri;
+
+  if (redirectUri && !is.uri(redirectUri)) {
+    throw new InvalidRequestError('Invalid request: `redirect_uri` is not a valid URI');
+  }
+
+  return Promise.try(this.model.getClient, clientId)
+    .then(function(client) {
+      if (!client) {
+        throw new InvalidClientError('Invalid client: client credentials are invalid');
+      }
+
+      if (!client.grants) {
+        throw new InvalidClientError('Invalid client: missing client `grants`');
+      }
+
+      if (!_.contains(client.grants, 'authorization_code')) {
+        throw new UnauthorizedClientError('Unauthorized client: `grant_type` is invalid');
+      }
+
+      if (!client.redirectUri) {
+        throw new InvalidClientError('Invalid client: missing client `redirectUri`');
+      }
+
+      if (redirectUri && client.redirectUri !== redirectUri) {
```

Comment:
A client may often have >1 redirect uri (for different apps).  The library should either compare `redirectUri` to `c'ient.redirectUri` which is of type array, or pass the redirectUri into getClient.